### PR TITLE
Sweeping Refactor

### LINKED ISSUE #4: Make subcomputations cacheable
I propose refactoring the interface to Astro to cache computations and subcomputations.

I would like the following (in addition to the normal, uncached interfaces if you want to roll like that):

```
class astro.solar.Time {
  constructor(time_julian)
  get mean_anomaly
  get longitude
  get obliquity
  get right_ascension
  get declination
  get distance
  event(lat, lon) { return new Event(this, lat, lon) }
}
class astro.solar.Event {
  constructor(time, lat, lon)
  get elevation
  get azimuth
  get hour_angle
}
```

Each of these things should memoize the relevant fields created as part of the computation.

In this way, most of the work involved in computing these properties is cached, and those caches are reused as much as possible. While objects get created, I'm fairly sure that the object allocation is much cheaper than the tons and tons of trig that we do in our other libraries.

### PR DESCRIPTION
**Pull request by ghost**
The interface to `astro` has been bugging me for a long time. I spent the effort to clean it up and add some new stuff that we'll want in order to make `dsgm-unpacker` more efficient.

1.  *BREAKING:* All input and output is now in degrees and milliseconds. We convert to radians and J2000 dates internally, but users of the interface no longer have to know or care about any of that.
2.  *BREAKING:* I renamed `astro.lunar.altitude` to `astro.lunar.elevation`, in order to match `astro.solar.elevation`. (Lemme know if you prefer `altitude`. It seems both are in common use. I don't really care!)
3.  I added `astro.lunar.obliquity` because why not.
4.  *BREAKING:* I moved `astro.hour_angle` to `astro.solar.hour_angle`.
5.  *BREAKING:* I squirreled away `astro.greenwich_mean_sidereal_time` and `astro.local_sidereal_time` into the internals, so they're not part of the interface any more.
6.  *BREAKING:* Methods for mucking with Julian dates are now hidden in the internals.
7.   I added `astro.solar.position`, which takes a time and returns an object storing the sun's position at that time. (From there, you can also call `.observer(lat, lon)`, which gives you the sun's position relative to that observer.) These functions basically give you cache objects for all the relevant solar computations, which will be useful for speeding up `dsgm-unpacker`. Fix #4 
8.  I bumped our devDependencies.
9.  I bumped our Travis testing versions.

I thought about adding `astro.solar.latitude` as a joke but ended up not. If you want me to, I can.

### CODE REVIEW

File: lib/solar.js [hunk 1]
```diff
@@ -1,175 +1,232 @@
...
 }
 
-// https://en.wikipedia.org/wiki/Sunrise_equation#Mean_solar_noon
-function mean_solar_noon(t, lon) {
-  return J0 - lon / (2 * Math.PI) + julian_cycle(t, lon);
+
+// Display methods. Internally we work with Julian dates and radians.
+// Externally we work with millisecond timestamps and degrees.
+
+function longitude(ms) {
```
Comment (nonrenewable) [inline lib/solar.js]: These looked so similar to `lunar.js` that i was going to offer to factor them out (before i realized they're totally different 🤣 ) you made the modules too good 😉

**Review by nonrenewable-zz — APPROVED** [review]
The restructuring changes definitely make things clearer to read/use; based on my experience with astro i'm not willing to gander whether performance changed too much though 😅 👍 from me but @jareyes should also give this a go over since his context is much richer here!

File: lib/solar.js [hunk 2]
```diff
@@ -1,175 +1,232 @@
 "use strict";
-const astro = require("../index");
+const julian = require("./julian");
```
[the remark this answers is not in the document]
Comment (ironwallaby) [reply lib/solar.js]: I didn't add it... it was already there! Just sitting there... totally unused...

https://github.com/darkskyapp/astro/blob/71bac749336e9500cd84dc569042e9c54ffe5a37/lib/julian.js

[the remark this answers is not in the document]
Comment (nonrenewable) [reply lib/solar.js]: Oh my local copy of `astro` was waaaaaaay outdated 😳 carry on! 😂

**Comment by ironwallaby** [comment]
I still need to performance test this :3

One thing I'm thinking I'll do is change the units: latitude/longitude would be all-degrees-all-the-time, but other angles (azimuth, elevation angle, ecliptic longitude, etc.) should all be returned/input in radians. This is because they're always used for trigonometry purposes and the repeated conversions complicate external code.

**Comment by ironwallaby** [comment]
I made the following changes:

1. Output angles are all radians now. (This is because one often does additional trig work after getting output, and radians is much easier.)
2. I added `solar.latitude()` (always 0 :rofl:) and `lunar.distance()` for consistency.
3. I noticed there was an unused file `/lib/astro.js` sitting around. Nothing referenced it, so I nuked it.

**Comment by ironwallaby** [comment]
I added some more tests and I'm calling this "good enough" and releasing a major version update.

**Pull request merged**